pkg/sql/schemachanger/scplan/internal/scstage/build.go

```diff
@@ -503,7 +499,32 @@ func (sb stageBuilder) isUnmetInboundDep(de *scgraph.DepEdge) bool {
 	_, fromIsCandidate := sb.fulfilling[de.From()]
 	switch de.Kind() {
 
-	case scgraph.PreviousStagePrecedence, scgraph.PreviousTransactionPrecedence:
+	case scgraph.PreviousTransactionPrecedence:
+		if !fromIsFulfilled {
+			// If the source node has not already been fulfilled in an earlier
+			// stage, then we can't consider scheduling the destination node
+			// in the current stage.
+			return true
+		}
+		// At this point, the source node has been fulfilled before the current
+		// stage.
+		if sb.bs.currentPhase > scop.PreCommitPhase {
+			// Post-commit, transactions cannot have multiple stages, therefore
+			// the source node was fulfilled in an earlier transaction and this
+			// inbound dependency is met.
+			return false
+		}
+		// At this point, the current stage is in the statement transaction.
+		// The only way this inbound dependency can be met is if the source node
+		// status is the initial status before the schema change even began,
+		// in which case the node was fulfilled before the statement transaction.
+		idx, ok := sb.bc.targetToIdx[de.From().Target]
+		if !ok {
+			panic(errors.AssertionFailedf("unknown target %q", de.From().Target))
+		}
+		return de.From().CurrentStatus != sb.bc.initial[idx]
```

Comment:
@postamar I hope this is not too cold. I have some questions: This function essentially answers the question "is node `n` currently un-schedulable at this stage due to this inbound PreviousTransactionPrecedence edge?"

Your first half of the answer makes sense to me but the second, highlighted half does not. In particular, if `e.from()` is already fulfilled and we are in StatementPhase or PrecommitPhase, why do we need to consult whether `e.from()` is in its initial status to determine whether we return true of false here? 

In other words, if `de.from()` is currently *not* at its initial status, this means it must have transitioned in the StatementPhase of some previous statement, but how is that count as "previous transaction"? 
I want to know the precious definition of "previous transaction" in "PreviousTransactionPrecedence".

Reply:
I think you've answered your own question in your 3rd paragraph: it's precisely because we're still in the same transaction that the dependency cannot be met (we're not allowed to visit `de.To()` in the same transaction as `de.From()`).